opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,260 @@
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        if self._single_state_key:
+            value = {key: value}
+        elif not isinstance(value, dict):
+            raise ValueError(
+                "When the matrix database is configured with single_state_key=False, "
+                "the value passed must be a dict."
+            )
+
+        data = await self.get(state_key, check_dict=value)
+        if data is None:  # dict could be empty
+            _LOGGER.error("Error putting key into matrix room")
+            return
+
+        if {**data, **value} == data:
+            _LOGGER.error("Not updating matrix state, as content hasn't changed.")
+            return
+        else:
+            data = {**data, **value}
+
+        if self.is_room_encrypted and self.should_encrypt:
+            for k, v in value.items():
+                room_event = await self.opsdroid.send(
+                    GenericMatrixRoomEvent(
+                        target=self.room_id, content={k: v}, event_type=self._event_type
+                    )
+                )
+                data[k] = {"encrypted_val": room_event.event_id}
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id} with {data}")
+
+        await self.opsdroid.send(
+            MatrixStateEvent(
+                self._event_type,
+                content=data,
+                target=self.room_id,
+                connector=self.connector,
+                state_key=state_key,
+            )
+        )
+
+        return True
+
+    async def get(self, key, check_dict=None):
+        """Get a value from the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        _LOGGER.debug(
+            f"Getting {key} from matrix room {self.room_id} with state_key={state_key}."
+        )
+
+        data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error getting {key} from matrix room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if data.transport_response.status == 404:
+            data.content.clear()
+        data = data.content
+
+        _LOGGER.debug(f"Got {data} from state request.")
+
+        check = check_dict or data
+
+        for k, v in data.items():
```

Comment:
Aren't we now decrypting all events? Shouldn't we need to do something like this:
```
if self._single_state_key:
    data = {key: data[key]}
```
before this loop, so we only decrypt the keys we need if single state key is true?

Reply:
Actually we should detect if the key is not in the return here for single state key
```python
if self._single_state_key:
    if key not in data:
        return None
    data = {key: data[key]}
```